File: internal/gmail/timed_labels.go
package gmail

import (
	"context"
	"log"
	"time"
)

// TimedLabel defines a label with a name and max age for expiration
type TimedLabel struct {
	Name   string
	MaxAge string // duration key used in timedLabelCutoff
}

// TimedArchiveLabels are labels that trigger archiving after a delay
var TimedArchiveLabels = []TimedLabel{
	{"📥/1d", "1d"},
	{"📥/1w", "7d"},
	{"📥/1m", "30d"},
	{"📥/1y", "365d"},
}

// TimedDeleteLabels are labels that trigger deletion after a delay
var TimedDeleteLabels = []TimedLabel{
	{"🗑️/1d", "1d"},
	{"🗑️/1w", "7d"},
	{"🗑️/1m", "30d"},
	{"🗑️/1y", "365d"},
}

// ArchiveAfterReadLabel is applied to emails that should be archived once the user reads them.
const ArchiveAfterReadLabel = "📥/read"

// ProcessTimedLabels searches for emails with expired timed labels and archives/trashes them.
// Also processes the archive-after-read label.
func (c *Client) ProcessTimedLabels(ctx context.Context) error {
	// Process archive labels
	for _, tl := range TimedArchiveLabels {
		if err := c.processTimedLabel(ctx, tl.Name, tl.MaxAge, false); err != nil {
			log.Printf("Error processing timed archive label %s: %v", tl.Name, err)
			continue
		}
	}

	// Process delete labels
	for _, tl := range TimedDeleteLabels {
		if err := c.processTimedLabel(ctx, tl.Name, tl.MaxAge, true); err != nil {
			log.Printf("Error processing timed delete label %s: %v", tl.Name, err)
			continue
		}
	}

	// Process archive-after-read
	if err := c.processArchiveAfterRead(ctx); err != nil {
		log.Printf("Error processing archive-after-read label: %v", err)
	}

	return nil
}

func (c *Client) processArchiveAfterRead(ctx context.Context) error {
	labelID, err := c.GetLabelID(ctx, ArchiveAfterReadLabel)
	if err != nil {
		return nil // Label doesn't exist yet
	}

	// List messages with this label that are NOT unread (i.e., have been read)
	// Gmail: messages with the label but without the UNREAD label
	msgs, err := c.service.Users.Messages.List(c.userID).
		LabelIds(labelID).
		Q("-is:unread").
		Context(ctx).
		Do()
	if err != nil {
		return err
	}

	if msgs.Messages == nil {
		return nil
	}

	for _, msg := range msgs.Messages {
		// Remove the label and archive
		if err := c.RemoveLabels(ctx, msg.Id, []string{labelID}); err != nil {
			continue
		}
		if err := c.ArchiveMessage(ctx, msg.Id); err != nil {
			continue
		}
		log.Printf("Archived read message %s (📥/read)", msg.Id)
	}

	return nil
}

func (c *Client) processTimedLabel(ctx context.Context, labelName string, maxAge string, trash bool) error {
	labelID, err := c.GetLabelID(ctx, labelName)
	if err != nil {
		return nil // Label doesn't exist yet, nothing to process
	}

	// List messages with this label
	msgs, err := c.service.Users.Messages.List(c.userID).
		LabelIds(labelID).
		Context(ctx).
		Do()
	if err != nil {
		return err
	}

	if msgs.Messages == nil {
		return nil
	}

	// Calculate the cutoff timestamp based on maxAge
	cutoff := timedLabelCutoff(maxAge)

	for _, msg := range msgs.Messages {
		// Get message to check date
		full, err := c.service.Users.Messages.Get(c.userID, msg.Id).
			Format("minimal").
			Context(ctx).
			Do()
		if err != nil {
			continue
		}

		// InternalDate is milliseconds since epoch
		if full.InternalDate < cutoff {
			// Remove the timed label
			if err := c.RemoveLabels(ctx, msg.Id, []string{labelID}); err != nil {
				continue
			}

			if trash {
				if err := c.TrashMessage(ctx, msg.Id); err != nil {
					continue
				}
				log.Printf("Trashed message %s (timed label %s expired)", msg.Id, labelName)
			} else {
				if err := c.ArchiveMessage(ctx, msg.Id); err != nil {
					continue
				}
				log.Printf("Archived message %s (timed label %s expired)", msg.Id, labelName)
			}
		}
	}

	return nil
}

func timedLabelCutoff(maxAge string) int64 {
	now := time.Now()
	switch maxAge {
	case "1d":
		return now.Add(-24 * time.Hour).UnixMilli()
	case "7d":
		return now.Add(-7 * 24 * time.Hour).UnixMilli()
	case "30d":
		return now.Add(-30 * 24 * time.Hour).UnixMilli()
	case "365d":
		return now.Add(-365 * 24 * time.Hour).UnixMilli()
	default:
		return now.UnixMilli()
	}
}

File: cloudflare/src/jobs/timed-labels.ts
import type { Env } from '../types/env';
import { getAllActiveUsers, updateUserToken } from '../db/users';
import {
  getLabelId,
  listMessagesByLabel,
  getMessageMetadata,
  removeLabels,
  archiveMessage,
  trashMessage,
  refreshAccessToken,
} from '../services/gmail';

interface TimedLabel {
  name: string;
  maxAgeMs: number;
}

const ARCHIVE_LABELS: TimedLabel[] = [
  { name: '📥/1d', maxAgeMs: 1 * 24 * 60 * 60 * 1000 },
  { name: '📥/1w', maxAgeMs: 7 * 24 * 60 * 60 * 1000 },
  { name: '📥/1m', maxAgeMs: 30 * 24 * 60 * 60 * 1000 },
  { name: '📥/1y', maxAgeMs: 365 * 24 * 60 * 60 * 1000 },
];

const DELETE_LABELS: TimedLabel[] = [
  { name: '🗑️/1d', maxAgeMs: 1 * 24 * 60 * 60 * 1000 },
  { name: '🗑️/1w', maxAgeMs: 7 * 24 * 60 * 60 * 1000 },
  { name: '🗑️/1m', maxAgeMs: 30 * 24 * 60 * 60 * 1000 },
  { name: '🗑️/1y', maxAgeMs: 365 * 24 * 60 * 60 * 1000 },
];

export async function processTimedLabels(env: Env): Promise<void> {
  const users = await getAllActiveUsers(env.DB);

  for (const user of users) {
    try {
      let accessToken = user.accessToken;

      // Refresh token if expired
      if (new Date(user.tokenExpiry) <= new Date()) {
        try {
          const refreshed = await refreshAccessToken(
            env.GOOGLE_CLIENT_ID,
            env.GOOGLE_CLIENT_SECRET,
            user.refreshToken,
          );
          accessToken = refreshed.access_token;
          const newExpiry = new Date(Date.now() + refreshed.expires_in * 1000).toISOString();
          await updateUserToken(env.DB, user.id, accessToken, user.refreshToken, newExpiry);
        } catch (e) {
          console.error(`timed-labels: token refresh failed for ${user.email}:`, e);
          continue;
        }
      }

      // Process archive labels
      for (const tl of ARCHIVE_LABELS) {
        await processLabel(accessToken, tl, false);
      }

      // Process delete labels
      for (const tl of DELETE_LABELS) {
        await processLabel(accessToken, tl, true);
      }

      // Process archive-after-read
      await processArchiveAfterRead(accessToken);

      console.log(`timed-labels: processed for ${user.email}`);
    } catch (e) {
      console.error(`timed-labels: failed for ${user.email}:`, e);
    }
  }
}

const ARCHIVE_AFTER_READ_LABEL = '📥/read';

async function processArchiveAfterRead(accessToken: string): Promise<void> {
  const labelId = await getLabelId(accessToken, ARCHIVE_AFTER_READ_LABEL);
  if (!labelId) return;

  // List messages with this label that are NOT unread (i.e., have been read)
  const messages = await listMessagesByLabel(accessToken, labelId, '-is:unread');
  if (messages.length === 0) return;

  for (const msg of messages) {
    try {
      await removeLabels(accessToken, msg.id, [labelId]);
      await archiveMessage(accessToken, msg.id);
    } catch (e) {
      console.error(`timed-labels: failed to archive read message ${msg.id}:`, e);
    }
  }
}

async function processLabel(accessToken: string, tl: TimedLabel, trash: boolean): Promise<void> {
  // Find the label ID
  const labelId = await getLabelId(accessToken, tl.name);
  if (!labelId) return; // Label doesn't exist yet, nothing to process

  // List messages with this label
  const messages = await listMessagesByLabel(accessToken, labelId);
  if (messages.length === 0) return;

  const cutoff = Date.now() - tl.maxAgeMs;

  for (const msg of messages) {
    try {
      // Get message date
      const metadata = await getMessageMetadata(accessToken, msg.id);

      if (metadata.internalDate < cutoff) {
        // Remove the timed label first
        await removeLabels(accessToken, msg.id, [labelId]);

        if (trash) {
          await trashMessage(accessToken, msg.id);
        } else {
          await archiveMessage(accessToken, msg.id);
        }
      }
    } catch (e) {
      // Log and continue -- don't let one message failure block others
      console.error(`timed-labels: failed to process message ${msg.id}:`, e);
    }
  }
}
